backend/infrahub/actions/tasks.py

```
@@ -104,16 +205,96 @@ async def configure_action_rules(
+    To avoid performance issues, we need to analyse the generator and fetch the right relationships
+    """
+
+    targets_per_kind: dict[str, dict[str, InfrahubNode]] = defaultdict(dict)
+
+    for target in targets:
+        targets_per_kind[target["node"]["__typename"]][target["node"]["id"]] = None
+
+    for kind, values in targets_per_kind.items():
+        nodes = await client.filters(
+            kind=kind, branch=branch_name, ids=list(values.keys()), populate_store=False, order=Order(disable=True)
+        )
+        for node in nodes:
+            targets_per_kind[kind][node.id] = node
+
+    return targets_per_kind
+
+
+async def _run_generators(
     branch_name: str,
     node_ids: list[str],
     generator_definition_id: str,
-    service: InfrahubServices,
+    client: InfrahubClient,
+    context: InfrahubContext | None = None,
 ) -> None:
-    mutation = Mutation(
-        mutation="CoreGeneratorDefinitionRun",
-        input_data={"data": {"id": generator_definition_id, "nodes": node_ids}},
-        query={"ok": None},
+    """Fetch generator metadata and submit per-target runs.
+
+    Args:
+        branch_name: Branch on which to execute.
+        node_ids: Node IDs to run against (restricts selection if provided).
+        generator_definition_id: Generator definition to execute.
+        client: InfrahubClient to query additional data.
+        context: Execution context passed to downstream workflow submissions.
+
+    Returns:
+        None
+
+    Raises:
+        ValueError: If the generator definition is not found or none of the requested
+            targets are members of the target group.
+    """
+    response = await client.execute_graphql(
+        query=get_generator_run_query(definition_id=generator_definition_id, target_ids=node_ids).render(),
+        branch_name=branch_name,
     )
+    data = response[InfrahubKind.GENERATORDEFINITION]["edges"][0]["node"]
```

While it's unlikely to happen if the generator was deleted after triggering it this could cause a confusing error. Like below we could raise an error if we didn't have anything within `response[InfrahubKind.GENERATORDEFINITION]["edges"]`

---

Good point, I added a check to ensure that edges is not empty